File: src/Sentan.Core/Corpus.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace Sentan.Core
{
    public class Corpus
    {
        IDictionary<string, int> _tokens = new Dictionary<string, int>();

        public int EntryCount
        {
            get { return _tokens.Values.Sum(); }
        }

        public void Add(Document document)
        {
            foreach (var token in document)
            {
                if (_tokens.ContainsKey(token))
                {
                    _tokens[token]++;
                }
                else
                {
                    _tokens.Add(token, 1);
                }
            }
        }

        public void LoadFromDirectory(string directory)
        {
            foreach (var file in Directory.GetFiles(directory, "*.txt"))
            {
                foreach (var line in File.ReadAllLines(file))
                {
                    Add(new Document(line));
                }
            }
        }

        public int GetTokenCount(string token)
        {
            return _tokens.ContainsKey(token) ? _tokens[token] : 0;
        }
    }
}

File: rb/src/engine/corpus.rb
require "#{File.dirname(__FILE__)}/document"

class Corpus

  def initialize
    @tokens = {}
  end

  def entry_count
    @tokens.values.inject(0, :+)
  end

  def add document
    document.each_token { |token|
      @tokens[token] = 0 unless @tokens.has_key? token
      @tokens[token] += 1
    }
  end

  def load_from_directory directory
    Dir.glob("#{directory}/*.txt") { |entry|
      IO.foreach(entry){ |line|
        add Document.new(line)
      }
    }
  end

  def token_count token
    @tokens[token] || 0
  end
end
